### PR TITLE
Add difficulty tabs to mods list

### LINKED ISSUE #559: Add Difficulty Tabs to Mods List
**Is your feature request related to a problem? Please describe.**
Currently all mods are shown in the `/mods` datatable at the same time.

**Describe the solution you'd like**
The datatable should show mods from only 1 parent difficulty at a time. The user should be able to select between parent difficulties using tabs as per [the mockup](../blob/main/public/images/mockups/modsPage/mockup.png).

**Describe alternatives you've considered**
N/A. This is a required feature.

**Additional context**
This change will affect the strings passed to the `ListSelect` component for #556.

### PR DESCRIPTION
**Pull request by ShouvikGhosh2048**
Closes #559

### CODE REVIEW

File: prisma/seed.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ const seedRandomData = async () => {
     for (let i = 1; i <= 5; i++) {
         const difficulty = await prisma.difficulty.create({
             data: {
-                name: "ExampleParentDifficulty" + i,
+                name: "Difficulty" + i,
```
Comment (ShouvikGhosh2048) [inline prisma/seed.ts:18]: The parent difficulty name needs to be a prefix of the child difficulty name for filtering. I've made that change.

File: src/components/mods/modsTable.tsx [hunk 2]
```diff
@@ -16,16 +16,35 @@ import type { ModWithInfo } from "~/components/mods/types";
 
 const PAGE_SIZES = [5, 10, 15, 20, 25, 50, 100, 250, 500, 1000];
 const DEFAULT_PAGE_SIZE_INDEX = 1;
-
+const TAB_COLORS = ["#a7a7a7", "#b7c3dc", "#92a6c5", "#527da2", "#263972"];
```
Comment (ShouvikGhosh2048) [inline src/components/mods/modsTable.tsx:19]: I used the colors in the mockup - I think they need to be changed for better contrast.

Comment (otobot1) [reply src/components/mods/modsTable.tsx:19]: Agreed. Please tweak the colors.

**Comment by otobot1** [comment]
A few notes: 
1. I agree about the tab colors. Please pick some better ones.
2. Please change the difficulty column filtering. The parent difficulty filtering should be performed only by the tabs and not the difficulty column filter. The difficulty column filter should allow sorting by sub-difficulty - allow the user to select "Red", "Yellow", or "Green". I neglected to add this to the notes in #559 - sorry about that.
3. Please have the table header and table footer change color to match the currently active tab so that it's clear which tab is active. I'll delay working on #602 for now.

**Comment by ShouvikGhosh2048** [comment]
I've implemented (2) and (3). Could you choose the colors for (1)? (I'm not good with choosing colors 😅)

How will we handle mods with no difficulty ratings?

File: src/components/mods/modsTable.tsx [hunk 3]
```diff
@@ -121,7 +167,13 @@ export const ModsTable = ({ qualities, difficulties, modsWithInfo, isLoading }:
     const [mapCountRange, setMapCountRange] = useState<[number | undefined, number | undefined]>([undefined, undefined]);     //[min, max]
     const [selectedModTypes, setSelectedModTypes] = useState<ModType[]>([]);
     const [selectedQualities, setSelectedQualities] = useState<string[]>([]);
-    const [selectedParentDifficulties, setSelectedParentDifficulties] = useState<string[]>([]);
+    const [selectedChildDifficulties, setSelectedChildDifficulties] = useState<string[]>([]);
+
+    // Reset tab index and selected child difficulties if the difficulties change.
+    useEffect(() => {
```
Comment (ShouvikGhosh2048) [inline src/components/mods/modsTable.tsx:173]: This effect runs when `difficulties` loads for the first time as well as when the `difficulties` returned by the server changes (this only happens when we change the difficulties in the database).

**Comment by otobot1** [comment]
1. 👍 
2. 👍 
3. 👍 Follow-up item: It's hard to see the greyed out up/down sorting arrows on the various table header colors.

Good question. Unrated mods should be sorted based on the lowest canonical difficulty of their constituent map(s) - every map in the database will *always* have a cannonicalDifficultyId so this is a good fallback. The difficulty name string will be [this](../blob/main/src/consts/noRatingsFoundMessage.ts), which should be added to the allowed strings in the difficulty filter if and only if it applies to a mod on that tab. If there are no quality ratings, the same string will be used, and it should be added to the quality filter in those cases as well.

**Comment by otobot1** [comment]
Also, I think it might be confusing that only the difficulty filter resets between tabs. Please change it so that the difficulty filter only displays "Red", "Yellow", and "Green" and doesn't reset between tabs.

**Comment by ShouvikGhosh2048** [comment]
1) Why are we fetching the difficulties from the server - aren't they fixed? Or do you want to allow changing the possible difficulties in the database in the future. If so what will the format of the difficulty name be - will it always be `<parentDifficulty>` and `<parentDifficulty>: <childDifficulty>`?

2) So just to confirm, for a mod with no ratings, the lowest parent difficulty amongst it's maps will be the tab in which the mod lies? What if it has no maps?

**Comment by otobot1** [comment]
1. Yeah exactly. There will likely be more difficulties in the future (people keep making harder mods), so I figure it's easier if it lives in the database. The naming format will stay the same, and all difficulties will have the same set of children.
2. Correct. All mods will have at least 1 map. We may support non-map mods in the future if that's  a highly requested feature, but they would be a new type and kept in a new table in the db.

**Comment by ShouvikGhosh2048** [comment]
@otobot1

I've implemented the new features. You can take a look.

**Comment by otobot1** [comment]
I ran into some bugs when testing it. I'll dm you the details.

File: src/components/mods/modsTable.tsx [hunk 4]
```diff
@@ -111,8 +142,50 @@ export const ModsTable = ({ qualities, difficulties, modsWithInfo, isLoading }:
         [difficulties],
     );
 
+    const [currentTabIndex, setCurrentTabIndex] = useState(null as null | number);
```
Comment (otobot1) [inline src/components/mods/modsTable.tsx:145]: The better way to do this is to pass the optional type parameter to the useState function

File: src/components/mods/modsTable.tsx [hunk 5]
```diff
@@ -330,148 +436,159 @@ export const ModsTable = ({ qualities, difficulties, modsWithInfo, isLoading }:
...
-            striped
-            textSelectionDisabled
-            withColumnBorders
-            highlightOnHover
-            fetching={isLoading}
-            records={records}
-            idAccessor={(record) => record.id}
-            columns={[
+        <div>
+            <div className={classes.tabContainer}>
```
Comment (otobot1) [inline src/components/mods/modsTable.tsx:443]: It's better to use a fragment unless you actually *need* a div. This way the rendered html won't contain an unnecessary element. This is marginally more efficient, but more importantly it's more semantically correct which avoids some problems encountered by screen readers (accessibility stuff).

**Review by otobot1 — APPROVED** [review]
Looks good! I made a couple of very minor changes (see review comments) but other than that mostly just adjusted formatting a bit.

**Pull request merged**